### PR TITLE
Rover: SmartRTL attempts to save home even after armed

### PR DESCRIPTION
**Pull request by oneTimePad**
Tested in SITL and with build_all.sh. #7448

### CODE REVIEW

**Comment by magicrub** [comment]
this definitely sounds like something copter should have too!

**Comment by khancyr** [comment]
Hello, thanks for your contribution !
I think the idea is good but not the fix. We should better update the home position from the smart_rtl_update() and not hide it inside SmartRTL otherwise GCS won't be warn where it has been set. 
Something like : 
````
void Rover::smart_rtl_update()
{
    //  reset path if home wasn't set and became set
    if (home_is_set == HOME_UNSET && update_home_from_EKF()) {
        g2.smart_rtl.reset_path(true);
    } else {  // return otherwise
        return;
    }
    const bool save_position = hal.util->get_soft_armed() && (control_mode != &mode_smartrtl);
    mode_smartrtl.save_position(save_position);
}
````
Well, this need to modify update_home_from_EKF() to take into account return correctly. And change  AP_SmartRTL::deactivat to : 
`````
void AP_SmartRTL::deactivate(SRTL_Actions action, const char *reason)
{
    if (_active) {
        _active = false;
        log_action(action);
        gcs().send_text(MAV_SEVERITY_WARNING, "SmartRTL deactivated: %s", reason);
    }
}
`````

What do you think ?

File: libraries/AP_SmartRTL/AP_SmartRTL.cpp [hunk 1]
```diff
@@ -166,6 +166,19 @@ bool AP_SmartRTL::pop_point(Vector3f& point)
     return true;
 }
 
+//Sets home and resets path when position is ok, ahrs has origin, and already armed. Avoids bad position GCS message
+void AP_SmartRTL::save_home(bool position_ok)
```
Comment (muramura) [inline libraries/AP_SmartRTL/AP_SmartRTL.cpp]: The indent of this method is 2. 4 White space is a code style.

**Comment by oneTimePad** [comment]
@khancyr  Thanks! I like that method a lot better; however, it seems `update_home_from_EKF()` is a void function.

File: libraries/AP_SmartRTL/AP_SmartRTL.cpp (hunk 1 above)
Comment (oneTimePad) [reply libraries/AP_SmartRTL/AP_SmartRTL.cpp]: Ah sorry about that. Will fix

**Comment by oneTimePad** [comment]
@khancyr  I tried the method you posted, and I noticed there is a race ( unless I am missing something else you wanted me to do). 

`update_home_from_EKF` is called continuously from `Rover::ahrs_update`. In addition,  `update_home_from_EKF` calls `Rover::set_home_from_current_location`, which  calls `Rover::set_home` and sets `home_is_set` to `HOME_SET_AND_LOCKED` on success. It is  possible, and it happened (since `ahrs_update` is called at a higher frequency), that if `ahrs_update` sets the home before `smart_rtl_update`  does (`ahs_update`'s `update_home_from_EKF` succeeds before the one in `smart_rtl_update`)  then `AP_SmartRTL::reset_path` will never be called. This is  due to the first part of the if condition (`home_is_set` will no longer be equal to `HOME_UNSET` and `reset_path` will never be called.

Is it safe to remove `update_home_from_EKF` from `ahrs_update`, and assume no other function will set `home_is_set` before `smart_rtl_update` does? Would it be better to add a separate bool like `smart_rtl_home_set` in Rover.h?

**Comment by rmackay9** [comment]
I've had a look at the PR and I generally agree with @khancyr that we should try a different approach.

What we want is:

- as a minimum we want the smart-rtl home to be set to the current location when the vehicle's position first becomes good
- as a "nice to have" if the user ever sets the ahrs home to the current location we should reset the smart-rtl home to the current location as well.  If the user ever sets the home to a location that is not our current position, the smart-rtl home should not be touched.  This will lead to a mismatch between the ahrs home and smart-rtl home but that can't be helped for now at least.

I think this set of changes might accomplish the above:

- rename AP_SmartRTL::reset_path(xxx) to AP_SmartRTL::set_home(xxx)
- modify Rover::set_home_to_current_location() so that if it's "set_home" call succeeds, the AP_SmartRTL::set_home() is called.
- modify Copter::set_home_to_current_location_inflight() and Copter::set_home_to_current_location to do the same as above.
- in AP_SmartRTL class, add the _saved_home member as you have in this PR (maybe call it _home_saved).
- in AP_SmartRTL::set_home (previously called reset_path) remove the deactivate and add a "_home_saved = true;"
- in AP_SmartRTL::update(), if position_ok but !_home_saved then call AP_SmartRTL::set_home.

I think this will result in smart-rtl saving a home position when the EKF first gets a good position whether it's armed or not at that moment.

If it wasn't armed when that smart-rtl home was set, it will actually start recording a path (say the users carries the vehicle around) but that path will be thrown away when the vehicle is armed so it shouldn't be a problem.  There is a small risk that if the user carries the vehicle a really long way that smart-rtl will be deactivated before the vehicle is armed but I think it's a small risk and smart-rtl will be reactivated when it is next armed.

If the vehicle was armed when smart-rtl home was set, it should start recording a path from that moment onwards and so it will return to that home position correctly.

Additionally if the user ever calls set-home-to-current-location from the ground station, that will now work but it wouldn't have worked before.

A weird edge case is if the user uses the ground station to set the ahrs home position to a location other than the vehicle's current location while the vehicle is disarmed.  This will mean the smart-rtl set-home is not called when the vehicle is armed and it will instead use whatever path it has built up from first GPS lock.  We can probably ignore this edge case or send a warning message to the user.

**Comment by oneTimePad** [comment]
@rmackay9  No problem, and thank you for your advice! I will get to work on these changes.

**Comment by khancyr** [comment]
@oneTimePad, Yep my solution got on problem on that, nice find !
The main point, contrary to what your are proposing, is that SmartRTL is mode and not the vehicle main code, so all the code shouldn't rely on smartrtl . It need to be SmartRTL that is adapted to what the main code is doing ! 
@rmackay9 nice solution !

**Comment by oneTimePad** [comment]
@rmackay9 So I'm slightly confused on one part; the last edge case. So as of now `Rover::arm_motors` calls `reset_path` , now renamed  `set_home`, unconditionally, is that how it still should be? So it seems `set_home` would be called in the last edge case, when armed. Unless I'm misunderstanding something. However, Copter doesn't seem to call `set_home` for SmartRTL home position when armed. I'm guessing that should be added in either `Copter::arm_motors_check` or `Copter::smart_rtl_update`.

I also noticed something about Copter and Rover that might now need to be changed:
`Rover::smart_rtl_update` only saves position (`ModeSmartRTL::save_position`) when `hal.util->get_soft_armed() && (control_mode != &mode_smartrtl)`, guessing that should be changed to just `control_mode != mode_smartrtl`. Since we want the path to be saved even if the user carries the vehicle disarmed as you mentioned. In addition, Copter calls `Copter::ModeSmartRTL::save_position` in `Copter::smart_rtl_save_position` (the equivalent of `Rover::smart_rtl_update`) unconditionally. It seems Copter should have the condition `(control_mode != &mode_smartrtl)` as well.

Just want to confirm these changes before I proceed to make sure I'm on the right track. I already made the other changes you specified above.

**Comment by rmackay9** [comment]
@oneTimePad,

I think the set_home should be called from within AP_SmartRTL::update() if the caller (i.e. Rover or Copter) starts setting the "position_ok" to true but has never called the AP_SmartRTL::set_home() function before.
>> in AP_SmartRTL::update(), if position_ok but !_home_saved then call AP_SmartRTL::set_home.

In some ways it's slightly bad behaviour to have different layers calling this set-home method because it makes it less clear who is responsible for setting home.. but we could also consider it a bit of a safety measure... the library steps in and saves home if the caller doesn't for some reason.

I agree that we should make the calls to AP_SmartRTL::update consistent in Rover and Copter.  It might make more sense to add the get-soft-armed check into the Copter version.  In general I don't think we *want* to store the path when the vehicle is being carried around although it's not particularly dangerous for it to do that.

**Comment by rmackay9** [comment]
Merged!  Really great work, thanks!

**Pull request closed without merging**